### PR TITLE
Add cross-cutting (=apteeraus) as a post-processing option

### PR DESCRIPTION
**Pull request by unie31**
closing #22.

This is an initial PR for cross cutting, hoping to hear improvement suggestions. I have already added some comments to the code, which may be interesting to you.

### CODE REVIEW

File: forestry/cross_cutting.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        volumes, values = _cross_cut(tree, r)
+
+        #NOTE: the above 'volumes' and 'values' are calculated for a single reference tree. To report meaningful numbers,
+        # they must be multiplied by the reference tree's stem count per ha and the stand area (in hectares)
+        multiplier = tree.stems_per_ha * (stand.area/1000) #area is given in square meters, thus need to convert to ha.
+        volumes = [vol*multiplier for vol in volumes] 
+        values = [val*multiplier for val in values]
+
+        volumes_bucket.append(volumes)
+        values_bucket.append(values)
```
Comment (unie31) [inline forestry/cross_cutting.py:51]: Note the comment: "the above 'volumes' and 'values' are calculated for a single reference tree. To report meaningful numbers, they must be multiplied by the reference tree's stem count per ha and the stand area (in hectares)."

File: forestry/operations.py [hunk 2]
```diff
@@ -39,6 +41,35 @@ def report_volume(payload: Tuple[ForestStand, dict], **operation_parameters) ->
     return stand, new_simulation_aggregates
 
 
+def cross_cut(payload: OperationPayload, **operation_parameters) -> OperationPayload:
```
[the remark this answers is not in the document]
Comment (jpy-luke) [reply forestry/operations.py:44]: No, don't take in OperationPayload. This is handled only in the `processor` function wrapper, which passes only the tuple[ForestStand, dict] for operation functions.

File: forestry/operations.py [hunk 3]
```diff
@@ -39,6 +41,35 @@ def report_volume(payload: Tuple[ForestStand, dict], **operation_parameters) ->
...
+    """
+    stand = payload.simulation_state
+    simulation_aggregates = payload.aggregated_results
+
+    # latest_aggregate = get_latest_operation_aggregate(simulation_aggregates, 'report_cross_cutting')
+    volumes, values = cross_cut_stand(stand)
+
+    total_volume, total_value = calculate_cross_cut_aggregates(volumes, values)
+
+    # need to multiply volumes and values by ref tree stems per ha and stand area to get total values and vols!
```
Comment (jpy-luke) [inline forestry/operations.py:56]: If it is needed, then do it. If there's uncertainty about it, ask Reijo.

File: forestry/cross_cutting.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    result = convert_r_named_list_to_py_dict(result)
+    return (result["volumes"], result["values"])
+
+
+def cross_cut_stand(stand: ForestStand) -> tuple[List[float], List[float]]:
+    """
+    Calculates the total volume and value of cross cutting in the :stand:. 
+    """
+    r = robjects.R() #initialise the singleton instance here so that the main function can be sourced only once and not for each tree
+    r.source("r/cross_cutting/cross_cutting_main.R")
```
Comment (jpy-luke) [inline forestry/cross_cutting.py:32]: We don't have a general system for loading R modules yet. In the mean time, do the dirty stateful solution I did for lmfor_volume loading, to ensure that the R script doesn't get intitialized multiple times on every call to this function.

File: forestry/cross_cutting.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+def _cross_cut(tree: ReferenceTree, r: robjects.R ) -> dict:
+    species = _cross_cut_species_mapper.get(tree.species, "birch") #birch is used as the default species in cross cutting
+    result = r["cross_cut"](
+        species,
+        round(tree.breast_height_diameter), #TODO: find a better way than rounding. cross_cut function currently doesn't handle floats well..
+        round(tree.height)
+    )
+    
+    result = convert_r_named_list_to_py_dict(result)
+    return (result["volumes"], result["values"])
```
Comment (unie31) [inline forestry/cross_cutting.py:24]: I used cProfile and snakeViz to run profiling on post_processing with two stands as input, and the result doesn't look too good. Maybe a bug in my implementation? Maybe rpy2 is slow?
 
image

Comment (jpy-luke) [reply forestry/cross_cutting.py:24]: Not surprising, but let's deal with it separately.

File: forestry/cross_cutting.py (hunk 1 above)
Comment (unie31) [reply forestry/cross_cutting.py:51]: Mehtätalo to be consulted about this

Comment (jpy-luke) [reply forestry/cross_cutting.py:54]: I will handle this after the vacation period. For now let's just keep take this as is.

**Review by jpy-luke — APPROVED** [review]

**Pull request merged**